<!DOCTYPE html>
<html lang="en"><head>
<script src="index_files/libs/clipboard/clipboard.min.js"></script>
<script src="index_files/libs/quarto-html/tabby.min.js"></script>
<script src="index_files/libs/quarto-html/popper.min.js"></script>
<script src="index_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="index_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="index_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="index_files/libs/quarto-html/quarto-syntax-highlighting-a6e161b2431e1f94a14e0f5d32135a3c.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.8.24">

  <meta name="author" content="Matheus Pimenta">
  <meta name="dcterms.date" content="2025-09-13">
  <title>Phylogenetics</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="index_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="index_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    html { -webkit-text-size-adjust: 100%; }
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
      }
    pre.numberSource { margin-left: 3em;  padding-left: 4px; }
    div.sourceCode
      { color: #24292e;  }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #24292e; } /* Normal */
    code span.al { color: #ff5555; font-weight: bold; } /* Alert */
    code span.an { color: #6a737d; } /* Annotation */
    code span.at { color: #d73a49; } /* Attribute */
    code span.bn { color: #005cc5; } /* BaseN */
    code span.bu { color: #d73a49; } /* BuiltIn */
    code span.cf { color: #d73a49; } /* ControlFlow */
    code span.ch { color: #032f62; } /* Char */
    code span.cn { color: #005cc5; } /* Constant */
    code span.co { color: #6a737d; } /* Comment */
    code span.cv { color: #6a737d; } /* CommentVar */
    code span.do { color: #6a737d; } /* Documentation */
    code span.dt { color: #d73a49; } /* DataType */
    code span.dv { color: #005cc5; } /* DecVal */
    code span.er { color: #ff5555; text-decoration: underline; } /* Error */
    code span.ex { color: #d73a49; font-weight: bold; } /* Extension */
    code span.fl { color: #005cc5; } /* Float */
    code span.fu { color: #6f42c1; } /* Function */
    code span.im { color: #032f62; } /* Import */
    code span.in { color: #6a737d; } /* Information */
    code span.kw { color: #d73a49; } /* Keyword */
    code span.op { color: #24292e; } /* Operator */
    code span.ot { color: #6f42c1; } /* Other */
    code span.pp { color: #d73a49; } /* Preprocessor */
    code span.re { color: #6a737d; } /* RegionMarker */
    code span.sc { color: #005cc5; } /* SpecialChar */
    code span.ss { color: #032f62; } /* SpecialString */
    code span.st { color: #032f62; } /* String */
    code span.va { color: #e36209; } /* Variable */
    code span.vs { color: #032f62; } /* VerbatimString */
    code span.wa { color: #ff5555; } /* Warning */
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="index_files/libs/revealjs/dist/theme/quarto-5ae9c59b2727fcd5357d4f034d95f9e0.css">
  <link href="index_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="index_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="index_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="index_files/libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="index_files/libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="index_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Phylogenetics</h1>
  <p class="subtitle">Alignment-free Methods</p>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Matheus Pimenta <a href="https://orcid.org/0000-0003-0698-6360" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a>
</div>
<div class="quarto-title-author-email">
<a href="mailto:omatheuspimenta@outlook.com">omatheuspimenta@outlook.com</a>
</div>
        <p class="quarto-title-affiliation">
            Luiz de Queiroz College of Agriculture, University of São Paulo
          </p>
    </div>
</div>

  <p class="date">September 13, 2025</p>
</section><section id="TOC">
<nav role="doc-toc"> 
<h2 id="toc-title">Table of Contents</h2>
<ul>
<li><a href="#/phylogeny-basic-concepts" id="/toc-phylogeny-basic-concepts">Phylogeny – Basic Concepts</a></li>
<li><a href="#/approaches" id="/toc-approaches">Approaches</a></li>
<li><a href="#/examples-of-alignment-free-tools" id="/toc-examples-of-alignment-free-tools">Examples of Alignment-Free Tools</a></li>
<li><a href="#/applications" id="/toc-applications">Applications</a></li>
<li><a href="#/conclusion" id="/toc-conclusion">Conclusion</a></li>
</ul>
</nav>
</section>
<section>
<section id="phylogeny-basic-concepts" class="title-slide slide level1 center">
<h1>Phylogeny – Basic Concepts</h1>

</section>
<section id="definitions-and-brief-history---i" class="slide level2">
<h2>Definitions and brief history - I</h2>
<div class="columns">
<div class="column" style="width:55%;">
<ul>
<li class="fragment"><p>Phylogeny: a hypothesis of evolutionary relationships, not an absolute truth</p></li>
<li class="fragment"><p>Historical contributions:</p>
<ul>
<li class="fragment">Darwin → descent with modification<br>
</li>
<li class="fragment">Hennig → cladistics and systematics</li>
</ul></li>
<li class="fragment"><p>Anatomy of a phylogenetic tree: root, nodes, branches, terminals, clades, polytomies</p></li>
</ul>
</div><div class="column" style="width:45%;">
<ul>
<li><span class="fragment">
<figure>
<img src="https://training.galaxyproject.org/training-material/topics/evolution/tutorials/abc_intro_phylo/images/Darwin_tree.png" width="75%">
<figcaption style="font-size:0.5em; text-align:center;">
Darwin’s original sketch of evolutionary tree.
</figcaption>
</figure>
</span></li>
</ul>
</div></div>
</section>
<section id="definitions-and-brief-history---ii" class="slide level2">
<h2>Definitions and brief history - II</h2>
<div style="position: relative; width: 80%; height: 500px; margin: auto; text-align:center;">
<figure class="fragment fade-out" data-fragment-index="1" style="position:absolute; top:0; left:0; width:100%; margin:0;">
<a href="https://www.bioinformatics.nl/bif20306/chapter3.html" target="_blank"> <img src="images/sars-cov-2.svg" style="width:100%; height:auto;"> </a>
<figcaption style="font-size:1em; margin-top:0.5em;">
SARS-CoV-2 phylogenetic tree - Mar.&nbsp;2024
</figcaption>
</figure>
<figure class="fragment" data-fragment-index="1" style="position:absolute; top:0; left:0; width:100%; margin:0;">
<a href="https://nextstrain.org/mpox/all-clades" target="_blank"> <img src="images/mkpx_phylo.png" style="width:100%; height:auto;"> </a>
<figcaption style="font-size:1em; margin-top:0.5em;">
Monkeypox phylogenetic tree - Sep.&nbsp;2025
</figcaption>
</figure>
</div>
</section>
<section id="anatomy-of-a-phylogenetic-tree---i" class="slide level2">
<h2>Anatomy of a Phylogenetic Tree - I</h2>
<div style="text-align:center; width:100%; margin:auto;">
<figure>
<a href="https://www.bioinformatics.nl/bif20306/chapter3.html" target="_blank"> <img src="https://www.bioinformatics.nl/bif20306/_images/tree-types.png" style="width:90%; height:auto;"> </a>
<figcaption style="font-size:0.5em; margin-top:0.5em;">
A rooted phylogenetic tree with its main parts and characteristics indicated. Here, the OTUs are GenBank plant chloroplast gene accessions, the names of which have been condensed. For the same data, the tree is given as additive tree (top) and as an ultrametric tree (bottom left) with branch lengths corresponding to time. On the right, flipped, the same tree as cladogram, with branch lengths only indicating the structure of the trees.
</figcaption>
</figure>
</div>
</section>
<section id="anatomy-of-a-phylogenetic-tree---ii" class="slide level2">
<h2>Anatomy of a Phylogenetic Tree - II</h2>
<div style="text-align:center; width:100%; margin:auto;">
<figure>
<a href="https://www.geneious.com/guides/understanding-phylogenetics" target="_blank"> <img src="https://images.ctfassets.net/ifq1osqu2ba1/5KBdd5VPTNEo1JZYhuDaKp/a986de45a104247ac162e0defbd35f68/GN_Phylogenetic_tree_learn2.jpg?w=770" style="width:90%; height:auto;"> </a>
<figcaption style="font-size:0.5em; margin-top:0.5em;">
<p>Different components of a phylogenetic tree.</p>
</figcaption>
</figure>
</div>
</section>
<section id="anatomy-of-a-phylogenetic-tree---iii" class="slide level2">
<h2>Anatomy of a Phylogenetic Tree - III</h2>
<div style="text-align:center; width:100%; margin:auto;">
<figure>
<a href="https://www.geneious.com/guides/understanding-phylogenetics" target="_blank"> <img src="https://images.ctfassets.net/ifq1osqu2ba1/7pA09dcmTU5qUHnHeSsWQZ/c7d52c01689c138a4f1c1074f6e2cad3/GN_Phylogenetic_trees-learn6.jpg" style="width:90%; height:auto;"> </a>
<figcaption style="font-size:0.5em; margin-top:0.5em;">
<p>Rooted tree (left) and an unrooted tree (right).</p>
</figcaption>
</figure>
</div>
</section>
<section id="why-infer-phylogenetic-relationships" class="slide level2">
<h2>Why Infer Phylogenetic Relationships?</h2>
<p>Aside from gaining a fundamental understanding of biology, other reasons include:</p>
<ul>
<li class="fragment">Designing vaccines, for example for SARS-CoV-2 and influenza</li>
<li class="fragment">Measuring phylogenetic diversity for guiding conservation efforts</li>
<li class="fragment">Understanding coevolution; for example, around 70% of emergent human diseases have come from other species</li>
<li class="fragment">Dating major evolutionary events to study the effects of environmental change on different species</li>
</ul>
</section>
<section id="classical-paradigm" class="slide level2">
<h2>Classical Paradigm</h2>
<ul>
<li>Orthologs and paralogs
<ul>
<li><div style="text-align:center; width:100%; margin:auto;">
<figure>
<a href="https://www.nlm.nih.gov/ncbi/workshops/2023-08_BLAST_evol/ortho_para.html" target="_blank"> <img src="https://www.nlm.nih.gov/ncbi/workshops/2023-08_BLAST_evol/img/ortholog_paralog.jpg" style="width:62%; height:auto;"> </a>
<figcaption style="font-size:0.5em; margin-top:0.5em;">
Gene 1 in the ancestral species undergoes a duplication event generating Gene 1a and Gene 1b. The ancestral species splits into two species, each with its own copy of Gene 1a and Gene 1b. Gene 1a in species one is the ortholog of Gene 1a in species two; Gene 1a and Gene 1b are paralogs; All four genes are homologs.
</figcaption>
</figure>
</div></li>
</ul></li>
</ul>
</section>
<section id="classical-paradigm-1" class="slide level2">
<h2>Classical Paradigm</h2>
<ul>
<li class="fragment">Orthologs and paralogs
<ul>
<li class="fragment"><div style="text-align:center; width:100%; margin:auto;">
<figure>
<a href="https://www.bioinformatics.nl/bif20306/chapter3.html" target="_blank"> <img src="https://www.bioinformatics.nl/bif20306/_images/ortho-para.svg" style="width:90%; height:auto;"> </a>
<figcaption style="font-size:0.5em; margin-top:0.5em;">
The challenge of paralogs: (A) Paralogous genes are created by gene duplication events. Gene X is duplicated in a (recent) common ancestor (RCA) of species A and B resulting in paralogous genes X and X’. Species A and B inherit both copies of the gene (unless one or the other is lost somewhere along the way). (B) Phylogenetic analysis of the X/X’ gene family gives two parallel phylogenies. All sequences of gene X are orthologues of each other, as are all sequences of gene X’. However, X and X’ are paralogues. Both the X and X’ subtrees show the true relationships among the three species. The subtrees are also each other’s natural outgroup, and as a result each subtree is rooted with the other (reciprocally rooting). (C ) A tree of the X/X’ gene family can be misleading if not all the sequences are included (because of incomplete sampling or gene loss). If the broken branches are missing, then the true species relationships are misrepresented.
</figcaption>
</figure>
</div></li>
</ul></li>
</ul>
</section>
<section id="classical-paradigm-2" class="slide level2">
<h2>Classical Paradigm</h2>
<ul>
<li>Orthologs and paralogs</li>
<li>Multiple Sequence Alignment (MSA) as a hypothesis of homology</li>
</ul>
<ul>
<li class="fragment">Post-alignment methods:
<ul>
<li class="fragment">Character-based: Parsimony, Maximum Likelihood, Bayesian inference</li>
<li class="fragment">Distance-based: Neighbor-Joining, UPGMA<br>
</li>
</ul></li>
</ul>
<ul>
<li><figure class="fragment" style="text-align:center; width:100%; margin:auto;">
<a href="https://www.geneious.com/guides/understanding-phylogenetics" target="_blank"> <img src="https://images.ctfassets.net/ifq1osqu2ba1/4JvaUzsS7fJchTXm3MdHjP/885abd115d5a67d8fc4ef807a5b20c98/GN_Phylogenetic_tree_Construction-learn7.png" style="width:60%; height:auto;"> </a>
<figcaption style="font-size:0.5em; margin-top:0.5em;">
Types of tree construction methods.
</figcaption>
</figure></li>
</ul>
</section>
<section id="classical-paradigm-3" class="slide level2">
<h2>Classical Paradigm</h2>
<ul>
<li>Orthologs and paralogs</li>
<li>Multiple Sequence Alignment (MSA) as a hypothesis of homology</li>
<li>Post-alignment methods:
<ul>
<li>Character-based: Parsimony, Maximum Likelihood, Bayesian inference</li>
<li>Distance-based: Neighbor-Joining, UPGMA</li>
</ul></li>
</ul>
<ul>
<li class="fragment">Limitations of MSA:
<ul>
<li class="fragment">High computational cost</li>
<li class="fragment">Parameter dependence (gaps, scoring matrices)</li>
<li class="fragment">Loss of collinearity due to rearrangements and horizontal gene transfer (HGT)</li>
</ul></li>
<li class="fragment">Phylogenetic trees are not directly observed but <em>inferred</em></li>
</ul>
</section></section>
<section>
<section id="approaches" class="title-slide slide level1 center">
<h1>Approaches</h1>

</section>
<section id="alignment-based-methods---based-in-msa" class="slide level2">
<h2>Alignment-Based Methods - Based in MSA</h2>
<div class="columns">
<div class="column" style="width:55%;">
<ul>
<li class="fragment">Uses a set of genes</li>
<li class="fragment">Character based
<ul>
<li class="fragment">maximum parsimony (MP)</li>
<li class="fragment">maximum likelihood (ML)</li>
<li class="fragment">Bayesian Inference (BI)</li>
</ul></li>
</ul>
</div><div class="column" style="width:45%;">
<ul>
<li><span class="fragment">
<figure>
<a href="https://www.bioinformatics.nl/bif20306/chapter3.html" target="_blank"> <img src="https://www.bioinformatics.nl/bif20306/_images/bifurcating_alt.svg" width="100%"> </a><br>

<figcaption style="font-size:0.5em; text-align:center;">
Having to assess such large numbers of trees falls under the category of ‘NP complete’ problems which cannot be solved in a lifetime even with unlimited resources.
</figcaption>
</figure>
</span></li>
</ul>
</div></div>
</section>
<section id="maximum-parsimony-mp" class="slide level2">
<h2>Maximum Parsimony (MP)</h2>
<ul>
<li class="fragment"><p>Maximum parsimony seeks to provide the simplest explanation for a phenomenon.</p></li>
<li class="fragment"><p>In phylogenetics, this means building all possible trees and selecting the best tree based on the smallest number of evolutionary changes (e.g., insertions, deletions, substitutions) required to explain the relatedness between sequences.</p></li>
<li class="fragment"><figure class="fragment" style="text-align:center; width:100%; margin:auto;">
<p><a href="https://www.bioinformatics.nl/bif20306/chapter3.html" target="_blank"> <img src="images/parsimony.png" style="width:60%; height:auto;"> </a></p>
<figcaption style="font-size:0.5em; margin-top:0.5em;">
<p>Character state changes for all characters in the MSA shown left are indicated on the trees and exemplar syn- and autapomorphies are indicated. Note that character 6 is invariant and therefore does not contribute to any tree. Also note that each substitution occurring in the MSA results in one extra step on the tree.</p>
</figcaption>
</figure></li>
</ul>
</section>
<section id="maximum-likelihood-ml" class="slide level2">
<h2>Maximum Likelihood (ML)</h2>
<ul>
<li class="fragment"><p>The maximum likelihood model constructs all possible trees and selects the tree that most likely predicts the relatedness between sequences based on a specific evolutionary model.</p></li>
<li class="fragment"><p>These models take into account how likely different genetic changes occur over time.</p></li>
<li class="fragment"><p>For example, maximum likelihood models can account for:</p>
<ul>
<li class="fragment">The probability of different types of substitutions (A → G, C → T, etc.)<br>
</li>
<li class="fragment">The frequency of transition (purine → purine, pyrimidine → pyrimidine) vs.&nbsp;transversion (purine → pyrimidine, pyrimidine → purine)<br>
</li>
<li class="fragment">The rate of substitutions among different sites within a DNA sequence</li>
</ul></li>
</ul>
</section>
<section id="distance-based-clustering" class="slide level2">
<h2>Distance-Based Clustering</h2>
<ul>
<li class="fragment"><p>The other main approach to tree building is <strong>clustering</strong>, which is <strong>distance-based</strong>.</p></li>
<li class="fragment"><p>Widely used in several applications, for instance in visualizing BLAST searches as Neighbor Joining trees.</p></li>
<li class="fragment"><p>Distance-based means that instead of comparing one character at a time across all sequences in the MSA, <strong>pairwise comparisons of entire sequences</strong> are made for all possible sequence pairs.</p></li>
<li class="fragment"><p>Pairwise comparisons yield pairwise distances, which can be <strong>ultrametric</strong> or <strong>Euclidean</strong>.</p></li>
<li class="fragment"><p>The relation between <strong>Distance</strong> and <strong>Similarity</strong> matrices is:</p>
<ul>
<li class="fragment">Distance matrix → diagonals are all 0’s<br>
</li>
<li class="fragment">Similarity matrix → diagonals are all 1’s</li>
</ul></li>
</ul>
</section>
<section id="building-distance-trees" class="slide level2">
<h2>Building Distance Trees</h2>
<ul>
<li class="fragment"><p>Pairwise distances are calculated from the same MSA and used to build a <strong>distance tree</strong> (e.g., Neighbor Joining).</p></li>
<li class="fragment"><p>The MSA is not further used in the analysis; only the distance values are used.</p></li>
<li class="fragment"><p>Both parsimony trees and distance trees may have the same topology, but parsimony trees contain more information</p></li>
<li class="fragment"><p>If substitutions accumulate in a clock-like manner, distances may be ultrametric, meaning matrix distances match tree distances.</p></li>
<li class="fragment"><p>In real data, distances often differ from tree distances, producing additive trees with unequal sister branch lengths.</p></li>
<li class="fragment"><p>In additive trees, the most similar sequences in the matrix may not be the most closely related on the tree.</p></li>
</ul>
</section>
<section id="neighbor-joining-nj-method" class="slide level2">
<h2>Neighbor Joining (NJ) – Method</h2>
<ul>
<li class="fragment"><p>Neighbor Joining is one of the most commonly used distance tree building methods, fast and effective for large MSAs (hundreds of sequences).</p></li>
<li class="fragment"><p>NJ starts with a fully unresolved tree containing all sequences and calculates the total tree length by summing all pairwise distances.</p></li>
<li class="fragment"><p>A pair of sequences is chosen and joined to start a small cluster (“neighbors”), and the total tree length is updated.</p></li>
<li class="fragment"><p>This step is repeated iteratively until all sequences are joined, minimizing the overall tree length at each step.</p></li>
<li class="fragment"><p>NJ produces unrooted trees; if needed, an outgroup can be used for rooting.</p></li>
</ul>
</section>
<section id="neighbor-joining-nj-properties-and-applications" class="slide level2">
<h2>Neighbor Joining (NJ) – Properties and Applications</h2>
<ul>
<li class="fragment"><p>No molecular clock assumption, allowing branch lengths to differ between neighbors (sisters).</p></li>
<li class="fragment"><p>Can generate trees with hundreds of sequences very quickly, making it useful for rapid assessment of phylogenetic structure.</p></li>
<li class="fragment"><p>NJ is a <em>clustering method</em>, grouping sequences by overall similarity, not by shared ancestry (synapomorphies).</p></li>
<li class="fragment"><p>For detailed phylogenetic studies, character-based methods are preferred, and NJ can be used in addition to check for possible incongruencies.</p></li>
<li class="fragment"><p>If incongruencies are found, it may indicate non-metric data, which could warrant additional methods.</p></li>
</ul>
</section>
<section id="alignment-free-methods" class="slide level2">
<h2>Alignment-Free Methods</h2>
<ul>
<li class="fragment"><p>Principle: from positional homology → to compositional signatures</p></li>
<li class="fragment"><p>Advantages: speed, scalability, robustness to rearrangements</p></li>
<li class="fragment"><p>Trade-off: loss of positional information</p></li>
</ul>
</section>
<section id="main-alignment-free-strategies" class="slide level2">
<h2>Main Alignment-Free Strategies</h2>
<ul>
<li class="fragment"><p>Two main categories of alignment-free methods are recognized:</p>
<ol type="1">
<li class="fragment">Word-based methods<br>
</li>
<li class="fragment">Information theory-based methods</li>
</ol></li>
<li class="fragment"><p>Word-based methods analyze the frequency and distribution of <strong>k-mers</strong> (fixed-length subsequences) extracted from the sequences, creating specific profiles for each genome.<br>
Examples: k-mers, Average Common Substring (ACS), strobmers, minimizers, …</p></li>
<li class="fragment"><p>Information theory-based methods quantify the <strong>complexity and structure</strong> of genomic sequences, identifying more complex patterns and evolutionary or functional relationships between different genomes.</p></li>
<li class="fragment"><p>Other methods: Chaos Game Representation, graph-based methods, vector representation like <code>sweep</code> method, mathematical transforms (Fourier, wavelets)</p></li>
</ul>
</section>
<section id="k-mers" class="slide level2">
<h2>k-mers</h2>
<ul>
<li><figure class="fragment" style="text-align:center; width:100%; margin:auto;">
<a href="https://kamimrcht.github.io/webpage/strobemers.html" target="_blank"> <img src="https://kamimrcht.github.io/webpage/files/kmer0.png" style="width:60%; height:auto;"> </a>
<figcaption style="font-size:0.5em; margin-top:0.5em;">
k-mers from two sequences
</figcaption>
</figure></li>
</ul>
</section>
<section id="information-theory-based" class="slide level2">
<h2>Information theory-based</h2>
<ul>
<li><figure class="fragment" style="text-align:center; width:100%; margin:auto;">
<a href="http://biorxiv.org/lookup/doi/10.1101/2022.06.13.495895" target="_blank"> <img src="images/entropy.png" style="width:60%; height:auto;"> </a>
<figcaption style="font-size:0.5em; margin-top:0.5em;">
Diagrammatic representation of the methodology to produce the IE spectrum.
</figcaption>
</figure></li>
</ul>
</section>
<section id="chaos-game-representation" class="slide level2">
<h2>Chaos Game Representation</h2>
<figure class="fragment" style="text-align:center; width:100%; margin:auto;">
<a href="https://www.researchgate.net/figure/Contribution-of-base-composition-to-chaos-game-representation-CGR-images-Top-row_fig3_12738224" target="_blank"> <img src="images/cgr.png" style="width:60%; height:auto;"> </a>
<figcaption style="font-size:0.5em; margin-top:0.5em;">
Example of Chaos Game Representation of whole genomes.
</figcaption>
</figure>
</section>
<section id="varkodes-chaos-game-representation-based-method" class="slide level2">
<h2><code>varKodes</code>: Chaos Game Representation based method</h2>
<figure class="fragment" style="text-align:center; width:100%; margin:auto;">
<a href="https://www.nature.com/articles/s41559-025-02752-1" target="_blank"> <img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41559-025-02752-1/MediaObjects/41559_2025_2752_Fig1_HTML.png?as=webp" style="width:41%; height:auto;"> </a>
<figcaption style="font-size:0.5em; margin-top:0.5em;">
varKodes method workflow based in the Chaos Game Representation
</figcaption>
</figure>
</section>
<section id="comparison-alignment-vs.-alignment-free" class="slide level2">
<h2>Comparison: Alignment vs.&nbsp;Alignment-Free</h2>
<table class="caption-top">
<colgroup>
<col style="width: 32%">
<col style="width: 29%">
<col style="width: 37%">
</colgroup>
<thead>
<tr class="header">
<th>Feature</th>
<th>Alignment-based</th>
<th>Alignment-free</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Speed</td>
<td>Slow</td>
<td>Very fast</td>
</tr>
<tr class="even">
<td>Scalability</td>
<td>Limited</td>
<td>High</td>
</tr>
<tr class="odd">
<td>Assumption</td>
<td>Positional homology</td>
<td>Compositional similarity</td>
</tr>
<tr class="even">
<td>Rearrangement robust.</td>
<td>Low</td>
<td>High</td>
</tr>
<tr class="odd">
<td>Resolution</td>
<td>Site-specific</td>
<td>Whole-genome</td>
</tr>
<tr class="even">
<td>Memory usage</td>
<td>High</td>
<td>Low</td>
</tr>
</tbody>
</table>
</section></section>
<section>
<section id="examples-of-alignment-free-tools" class="title-slide slide level1 center">
<h1>Examples of Alignment-Free Tools</h1>

</section>
<section id="selected-tools-to-the-class" class="slide level2">
<h2>Selected tools to the class</h2>
<ul>
<li class="fragment"><code>mash</code>: pairwise distance estimation from sketches</li>
<li class="fragment"><code>mashtree</code>: <code>minHash</code> rapid phylogenomic surveillance</li>
<li class="fragment"><code>mike</code>: MinHash-based k-mer algorithm</li>
</ul>
</section>
<section id="selected-tools-to-the-class-1" class="slide level2">
<h2>Selected tools to the class</h2>
<ul>
<li><code>mash</code>: pairwise distance estimation from sketches</li>
<li><code>mashtree</code>: <code>minHash</code> rapid phylogenomic surveillance</li>
<li><code>mike</code>: MinHash-based k-mer algorithm</li>
</ul>
<div title="Tip">
<div class="callout callout-tip callout-titled callout-style-default">
<div class="callout-body">
<div class="callout-title">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<p><strong>Tip</strong></p>
</div>
<div class="callout-content">
<ul>
<li>choice of <strong>k-mer size</strong> strongly impacts results</li>
</ul>
</div>
</div>
</div>
</div>
</section>
<section id="mash-fast-genome-and-metagenome-distance-estimation-using-minhash" class="slide level2">
<h2><code>mash</code>: fast genome and metagenome distance estimation using MinHash</h2>
<ul>
<li class="fragment">Problem:
<ul>
<li class="fragment">Comparing large genomes using traditional alignment is slow and computationally expensive.<br>
</li>
<li class="fragment">We need a faster way to estimate similarity between sequences.</li>
</ul></li>
<li class="fragment">Mash’s Solution: MinHash Sketching
<ol type="1">
<li class="fragment"><strong>Decomposition into k-mers:</strong> Break sequences into overlapping substrings of length <em>k</em>.<br>
</li>
<li class="fragment"><strong>Hashing:</strong> Convert each k-mer into a unique number.<br>
</li>
<li class="fragment"><strong>Create the Sketch:</strong> Keep only the smallest hashes (e.g., 1000 smallest).<br>
</li>
</ol></li>
</ul>
</section>
<section id="mash-from-sketches-to-distance" class="slide level2">
<h2><code>mash</code>: From Sketches to Distance</h2>
<div class="columns">
<div class="column" style="width:60%;">
<ul>
<li class="fragment">The sketch is a small, random, representative sample of the genome, allowing massive data compression.</li>
</ul>
<ol type="1">
<li class="fragment"><strong>Estimate Similarity (Jaccard Index)</strong>
<ul>
<li class="fragment">Compare sketches by calculating the Jaccard Index:<br>
<span class="math display">\[j = \frac{|SketchA \cap SketchB|}{|SketchA \cup SketchB|}\]</span></li>
<li class="fragment">Higher <em>j</em> → genomes are more similar.</li>
</ul></li>
</ol>
</div><div class="column" style="width:40%;">
<ul>
<li class="fragment"><a href="https://kamimrcht.github.io/webpage/sketch.html" target="_blank"> <img src="images/jacc.png" style="width:100%; height:auto;"> </a>
<figcaption style="font-size:0.8em; margin-top:0.5em;">
Jaccard Similarity visually represented
</figcaption></li>
</ul>
</div></div>
</section>
<section id="mash-from-sketches-to-distance-1" class="slide level2">
<h2><code>mash</code>: From Sketches to Distance</h2>
<ol start="2" type="1">
<li class="fragment">Calculate Mash Distance (D)<br>
</li>
</ol>
<ul>
<li class="fragment"><p>Converts similarity to evolutionary distance:<br>
<span class="math display">\[D = -\frac{1}{k} \ln \frac{2j}{1+j}\]</span></p></li>
<li class="fragment"><p>Fast approximation of mutation rate, correlates with alignment-based measures.</p></li>
<li class="fragment"><p>Mash is scalable, fast, and ideal for clustering genomes, identifying unknown samples, and large-scale comparative genomics.</p></li>
</ul>
</section>
<section id="mash" class="slide level2">
<h2><code>mash</code></h2>
<ul>
<li>Used in all sequences from NCBI RefSeq Release 70 totaling 54,118 organisms and 618 Gbp of genomic sequence.</li>
<li>Sketching all genomes and computing all ~1.5 billion pairwise distances required just 26.1 and 6.9 CPU h, respectively.</li>
<li><figure class="fragment" style="text-align:center; width:100%; margin:auto;">
<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0997-x" target="_blank"> <img src="https://media.springernature.com/full/springer-static/image/art%3A10.1186%2Fs13059-016-0997-x/MediaObjects/13059_2016_997_Fig3_HTML.gif?as=webp" style="width:41%; height:auto;"> </a>
<figcaption style="font-size:0.5em; margin-top:0.5em;">
Comparison and de novo clustering of all RefSeq genomes using Mash. Each graph node represents a genome. Two genomes are connected by an edge if their Mash distance D ≤0.05 and P value ≤10–10. Only components containing microbial genomes are shown here (including viruses).
</figcaption>
</figure></li>
</ul>
</section>
<section id="mashtree-a-rapid-comparison-of-whole-genome-sequence-files." class="slide level2">
<h2><code>mashtree</code>: a rapid comparison of whole genome sequence files.</h2>
<ol type="1">
<li class="fragment">MinHash Sketching (Mash)
<ul>
<li class="fragment"><code>mash</code> creates genome sketches (sketch size increased from 1,000 → 10,000 for higher resolution) and calculates pairwise distances → stored in a distance matrix.</li>
</ul></li>
<li class="fragment">Neighbor-Joining Tree (QuickTree)
<ul>
<li class="fragment">Distance matrix used with QuickTree (default options) → generates a dendrogram.</li>
</ul></li>
</ol>
<ul>
<li class="fragment">Confidence Values:
<ul>
<li class="fragment">Although Mashtree does not infer true phylogeny, node support values are calculated using resample methods.<br>
</li>
</ul></li>
<li class="fragment">Two resampling methods for creating trees and assigning confidence to internal nodes.</li>
</ul>
</section>
<section id="mashtree-a-rapid-comparison-of-whole-genome-sequence-files.-1" class="slide level2">
<h2><code>mashtree</code>: a rapid comparison of whole genome sequence files.</h2>
<figure class="fragment" style="text-align:center; width:100%; margin:auto;">
<a href="https://joss.theoj.org/papers/10.21105/joss.01762" target="_blank"> <img src="images/mashtree1.png" style="width:50%; height:auto;"> </a>
<figcaption style="font-size:0.5em; margin-top:0.5em;">
The Mashtree workflow.
</figcaption>
</figure>
</section>
<section id="mashtree-a-rapid-comparison-of-whole-genome-sequence-files.-2" class="slide level2">
<h2><code>mashtree</code>: a rapid comparison of whole genome sequence files.</h2>
<figure class="fragment" style="text-align:center; width:100%; margin:auto;">
<a href="https://joss.theoj.org/papers/10.21105/joss.01762" target="_blank"> <img src="images/mashtree2.png" style="width:55%; height:auto;"> </a>
<figcaption style="font-size:0.5em; margin-top:0.5em;">
The Mashtree workflow - bootstrap mode.
</figcaption>
</figure>
</section>
<section id="mike-an-ultrafast-assembly-and-alignment-free-approach-for-phylogenetic-tree-construction" class="slide level2">
<h2><code>mike</code>: an ultrafast, assembly, and alignment-free approach for phylogenetic tree construction</h2>
<p>Build phylogenetic trees directly from raw sequencing reads, without assembly or alignment.</p>
<ul>
<li class="fragment">Steps:
<ol type="1">
<li class="fragment">Split reads into k-mers (k = 21).<br>
</li>
<li class="fragment">Group by prefix (kpre), encode suffixes (ksuf) into one-hot vectors.<br>
</li>
<li class="fragment">Random permutation → smallest non-zero value → minhash.<br>
</li>
<li class="fragment">Combine fingerprints → compute Jaccard similarity → convert to evolutionary distance.<br>
</li>
</ol></li>
<li class="fragment">Key advantages:
<ul>
<li class="fragment"><span class="math inline">\(\mathcal{O}(n)\)</span> time complexity (linear), low memory, accurate even for polyploids</li>
</ul></li>
</ul>
</section>
<section id="mike-an-ultrafast-assembly-and-alignment-free-approach-for-phylogenetic-tree-construction-1" class="slide level2">
<h2><code>mike</code>: an ultrafast, assembly, and alignment-free approach for phylogenetic tree construction</h2>
<ul>
<li class="fragment">Accuracy:
<ul>
<li class="fragment">Best performance around 10×–20× coverage.<br>
</li>
<li class="fragment">Outperforms Mash, BinDash, and Kssd in most scenarios.<br>
</li>
</ul></li>
<li class="fragment">Efficiency:
<ul>
<li class="fragment">Shortest runtime among minhash-based tools.<br>
</li>
<li class="fragment">Memory ≈200 MB (much lower than Kssd at &gt;30 GB).<br>
</li>
</ul></li>
<li class="fragment">Tree construction:
<ul>
<li class="fragment">Uses Neighbor Joining (NJ/BIONJ) with evolutionary distances.<br>
</li>
<li class="fragment">Performs well across haploid, autotetraploid, and allotetraploid simulations.<br>
</li>
<li class="fragment">Robust for large datasets (e.g., Ficus, Zea, Oryza).<br>
</li>
</ul></li>
</ul>
</section>
<section id="mike-saccharum-results-polyploid-sugarcane" class="slide level2">
<h2><code>mike</code>: Saccharum Results (Polyploid Sugarcane)</h2>
<ul>
<li>Dataset: 42 <em>Saccharum spontaneum</em> individuals (ploidy n=4 to n=13), Sorghum as outgroup.<br>
</li>
<li>Findings:
<ul>
<li>Topology consistent with earlier studies.<br>
</li>
<li>Individuals cluster into 4 distinct groups:
<ul>
<li>Group I: Tetraploid (n=4) + part of hexaploid (n=6).<br>
</li>
<li>Group II: Remaining hexaploid + nonuploid (n=9).<br>
</li>
<li>Groups III–IV: Decaploid (n=10) and higher ploidy species.<br>
</li>
</ul></li>
</ul></li>
<li>Conclusion:
<ul>
<li>MIKE is effective for polyploid species like sugarcane.<br>
</li>
<li>Can distinguish lineage differences where assembly/alignment methods are slow or impractical.</li>
</ul></li>
</ul>
</section>
<section id="mike-saccharum-results-polyploid-sugarcane-1" class="slide level2">
<h2><code>mike</code>: Saccharum Results (Polyploid Sugarcane)</h2>
<figure class="fragment" style="text-align:center; width:100%; margin:auto;">
<a href="https://doi.org/10.1093/bioinformatics/btae154" target="_blank"> <img src="https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bioinformatics/40/4/10.1093_bioinformatics_btae154/1/btae154f5.jpeg?Expires=1760550086&amp;Signature=FnYyoyhNeKp34FOtJcymwEBOayR1LdOaPeBPgi-HvoXiYq15s3Q~3qH~EGrqGIJ90whxdoOJ8inGLE3eEbuQHvUUHPxXIfsRi0gucxwfK4zG3-F2DYqMgGb~OHAkiqoX3INMRKHMkQl7sqQN5m5RWclZDyI8H-gpBF2GzSzXPeXc5t4MiqPN83t5T9s~HlOUjlCwFORxRwfMsV~Yz3trduJzyhwMePIS2wYgjkKy93qZbglQoth50nK3AHcGvKib9Llfo40T---f9xrOBC9ApZz~UbYU0p2zWRTdYa5w56iNJAysQl82T8NudTqLmBfYbGkDrXK4cg2M02X59jurag__&amp;Key-Pair-Id=APKAIE5G5CRDK6RD3PGA" style="width:75%; height:auto;"> </a>
<figcaption style="font-size:0.5em; margin-top:0.5em;">
The phylogenetic tree constructed using mike for 42 samples of S.spontaneum
</figcaption>
</figure>
</section></section>
<section>
<section id="applications" class="title-slide slide level1 center">
<h1>Applications</h1>

</section>
<section id="example" class="slide level2">
<h2>Example</h2>
<p>All files and softwares are in the <code>phylo</code> folder.<br>
Monkeypox sequences are from <a href="https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/">NCBI</a><br>
HIV sequences are from <a href="https://www.hiv.lanl.gov/content/index">Los Alamos National Laboratory</a></p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb1"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href=""></a><span class="fu">git</span> clone https://github.com/omatheuspimenta/phylo.git</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="column" style="width:50%;">
<p>MKPX</p>
<ul>
<li>41 sequences</li>
<li>5 subtypes</li>
<li>Average sequence size <span class="math inline">\(\approx\)</span> 200000 bp</li>
</ul>
</div><div class="column" style="width:50%;">
<p>HIV</p>
<ul>
<li>6964 sequences</li>
<li>8 subtypes</li>
<li>Average sequence size <span class="math inline">\(\approx\)</span> 10000 bp</li>
</ul>
</div></section>
<section id="install-softwares" class="slide level2">
<h2>Install softwares</h2>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb2" data-code-line-numbers="|2|4|6-7"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href=""></a><span class="co"># To install mashtree (conda version)</span></span>
<span id="cb2-2"><a href=""></a><span class="ex">conda</span> create <span class="at">-n</span> mashtree <span class="at">-c</span> bioconda mashtree</span>
<span id="cb2-3"><a href=""></a><span class="co"># To install mike just clone the github repository</span></span>
<span id="cb2-4"><a href=""></a><span class="fu">git</span> clone https://github.com/Argonum-Clever2/mike.git</span>
<span id="cb2-5"><a href=""></a><span class="co"># To install all dependencies to reproduce this class execute the .sh files</span></span>
<span id="cb2-6"><a href=""></a><span class="ex">./install_miniconda.sh</span> <span class="co"># install miniconda</span></span>
<span id="cb2-7"><a href=""></a><span class="ex">./install_r_env.sh</span> <span class="co"># create r env in conda and install all R packages used in the examples</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="example---mike-in-mkpx" class="slide level2">
<h2>Example - <code>mike</code> in MKPX</h2>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb3" data-code-line-numbers="|2|4-6|8|10|12|14|16|18"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href=""></a><span class="co"># Create a absolute path for every file</span></span>
<span id="cb3-2"><a href=""></a><span class="fu">realpath</span> MKPX/seqs/<span class="pp">*</span> <span class="op">&gt;</span> seqs_MKPX.txt</span>
<span id="cb3-3"><a href=""></a><span class="co"># Run the KMC kmers sketch</span></span>
<span id="cb3-4"><a href=""></a><span class="cf">while</span> <span class="bu">read</span> <span class="va">seq</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb3-5"><a href=""></a>    <span class="ex">python</span> kmc.py <span class="at">-k</span> 21 <span class="at">-f</span> <span class="st">"</span><span class="va">$seq</span><span class="st">"</span> <span class="at">-d</span> ../MKPX/output_kmers/</span>
<span id="cb3-6"><a href=""></a><span class="cf">done</span> <span class="op">&lt;</span> seqs_MKPX.txt</span>
<span id="cb3-7"><a href=""></a><span class="co"># Create absolute path for every kmer file</span></span>
<span id="cb3-8"><a href=""></a><span class="fu">realpath</span> MKPX/output_kmers/<span class="pp">*</span> <span class="op">&gt;</span> kmers_MKPX.txt</span>
<span id="cb3-9"><a href=""></a><span class="co"># Run mike sketch -t 1 (thread number)</span></span>
<span id="cb3-10"><a href=""></a><span class="ex">./mike</span> sketch <span class="at">-t</span> 1 <span class="at">-l</span> kmers_MKPX.txt <span class="at">-d</span> MKPX/output_mike/</span>
<span id="cb3-11"><a href=""></a><span class="co"># Create absolute path for every sketch file</span></span>
<span id="cb3-12"><a href=""></a><span class="fu">realpath</span> MKPX/output_mike/<span class="pp">*</span> <span class="op">&gt;</span> mike_MKPX.txt</span>
<span id="cb3-13"><a href=""></a><span class="co"># Compute Jaccard (mike)</span></span>
<span id="cb3-14"><a href=""></a><span class="ex">./mike</span> compute <span class="at">-l</span> mike_MKPX.txt <span class="at">-L</span> mike_MKPX.txt <span class="at">-d</span> MKPX/</span>
<span id="cb3-15"><a href=""></a><span class="co"># Compute distance (mike)</span></span>
<span id="cb3-16"><a href=""></a><span class="ex">./mike</span> dist <span class="at">-l</span> mike_MKPX.txt <span class="at">-L</span> mike_MKPX.txt <span class="at">-d</span> MKPX/</span>
<span id="cb3-17"><a href=""></a><span class="co"># Create newick file with R</span></span>
<span id="cb3-18"><a href=""></a><span class="ex">Rscript</span> draw.r <span class="at">-f</span> MKPX/dist.txt <span class="at">-o</span> MKPX/dist.nwk</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="example---mashtree-in-mkpx" class="slide level2">
<h2>Example - <code>mashtree</code> in MKPX</h2>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb4" data-code-line-numbers="|2|4|6"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href=""></a><span class="co"># Activate conda environments</span></span>
<span id="cb4-2"><a href=""></a><span class="ex">conda</span> activate mashtree</span>
<span id="cb4-3"><a href=""></a><span class="co"># Run mashtree in fasta files</span></span>
<span id="cb4-4"><a href=""></a><span class="ex">mashtree</span> MKPX/seqs/<span class="pp">*</span>.fasta <span class="op">&gt;</span> MKPX/mashtree.dnd</span>
<span id="cb4-5"><a href=""></a><span class="co"># Run mashtree bootstrap mode (only 100 times in class, in pratice 1000)</span></span>
<span id="cb4-6"><a href=""></a><span class="ex">mashtree_bootstrap.pl</span> <span class="at">--reps</span> 100 MKPX/seqs/<span class="pp">*</span>.fasta <span class="at">--</span> <span class="at">--min-depth</span> 0 <span class="op">&gt;</span> MKPX/mashtree.bootstrap.dnd</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="plotting-the-trees" class="slide level2">
<h2>Plotting the Trees</h2>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb5" data-code-line-numbers="|2|4|6"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href=""></a><span class="co"># Plot mike output</span></span>
<span id="cb5-2"><a href=""></a><span class="ex">Rscript</span> create_tree_MKPX.R <span class="at">-t</span> MKPX/dist.nwk <span class="at">-a</span> MKPX/annotation.csv <span class="at">-o</span> MKPX/</span>
<span id="cb5-3"><a href=""></a><span class="co"># Plot mashtree output</span></span>
<span id="cb5-4"><a href=""></a><span class="ex">Rscript</span> create_tree_MKPX.R <span class="at">-t</span> MKPX/mashtree.dnd <span class="at">-a</span> MKPX/annotation.csv <span class="at">-o</span> MKPX/</span>
<span id="cb5-5"><a href=""></a><span class="co"># Plot mashtree bootstrap output</span></span>
<span id="cb5-6"><a href=""></a><span class="ex">Rscript</span> create_tree_MKPX.R <span class="at">-t</span> MKPX/mashtree.bootstrap.dnd <span class="at">-a</span> MKPX/annotation.csv <span class="at">-o</span> MKPX/</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="plots" class="slide level2">
<h2>Plots</h2>
<div style="position: relative; width: 80%; height: 500px; margin: auto; text-align:center;">
<!-- First image - visible initially, fades out when second appears -->
<figure class="fragment fade-out" data-fragment-index="1" style="position:absolute; top:0; left:0; width:100%; margin:0;">
<a href="blank" target="_blank"> <img src="images/dist_phylo_tree1.png" style="width:100%; height:auto;"> </a>
<figcaption style="font-size:1em; margin-top:0.5em;">
MKPX phylogenetic tree - mike version
</figcaption>
</figure>
<!-- Second image - appears at the same time first fades out -->
<figure class="fragment fade-in-then-out" data-fragment-index="1" style="position:absolute; top:0; left:0; width:100%; margin:0;">
<a href="blank" target="_blank"> <img src="images/mashtree_phylo_tree1.png" style="width:100%; height:auto;"> </a>
<figcaption style="font-size:1em; margin-top:0.5em;">
Monkeypox phylogenetic tree - mashtree version
</figcaption>
</figure>
<!-- Third image - appears when second fades out -->
<figure class="fragment fade-in" data-fragment-index="2" style="position:absolute; top:0; left:0; width:100%; margin:0;">
<a href="blank" target="_blank"> <img src="images/mashtree.bootstrap_phylo_tree1.png" style="width:100%; height:auto;"> </a>
<figcaption style="font-size:1em; margin-top:0.5em;">
Monkeypox phylogenetic tree - mashtree bootstrap version
</figcaption>
</figure>
</div>
</section>
<section id="example---mashtree-in-hiv-challenge" class="slide level2">
<h2>Example - <code>mashtree</code> in HIV (challenge)</h2>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb6" data-code-line-numbers="|2|4|6|8|10"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href=""></a><span class="co"># Activate conda environments</span></span>
<span id="cb6-2"><a href=""></a><span class="ex">conda</span> activate mashtree</span>
<span id="cb6-3"><a href=""></a><span class="co"># Use only 500 sequences to the class (run in the HIV folder)</span></span>
<span id="cb6-4"><a href=""></a><span class="ex">./select_500seqs.sh</span></span>
<span id="cb6-5"><a href=""></a><span class="co"># Copy the sequences into 'samples' folder</span></span>
<span id="cb6-6"><a href=""></a><span class="fu">mkdir</span> samples <span class="kw">&amp;&amp;</span> <span class="fu">cat</span> samples.txt <span class="kw">|</span> <span class="fu">xargs</span> cp <span class="at">-t</span> samples/</span>
<span id="cb6-7"><a href=""></a><span class="co"># Run mashtree in fasta files - 11-mers. (can use --numcpus 4 option to run in multithread mode)</span></span>
<span id="cb6-8"><a href=""></a><span class="ex">mashtree</span> <span class="at">--kmerlength</span> 11 HIV/samples/<span class="pp">*</span>.fasta <span class="op">&gt;</span> HIV/mashtree.dnd</span>
<span id="cb6-9"><a href=""></a><span class="co"># Run mashtree bootstrap mode (only 100 times in class, in pratice 1000)</span></span>
<span id="cb6-10"><a href=""></a><span class="ex">mashtree_bootstrap.pl</span> <span class="at">--reps</span> 100 HIV/samples/<span class="pp">*</span>.fasta <span class="at">--</span> <span class="at">--min-depth</span> 0 <span class="at">--kmerlength</span> 11 <span class="op">&gt;</span> HIV/mashtree.bootstrap.dnd</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="plotting-the-trees-1" class="slide level2">
<h2>Plotting the Trees</h2>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb7" data-code-line-numbers="|2|4"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href=""></a><span class="co"># Plot mashtree output</span></span>
<span id="cb7-2"><a href=""></a><span class="ex">Rscript</span> create_tree_HIV.R <span class="at">-t</span> HIV/mashtree.dnd <span class="at">-a</span> HIV/annotation.csv <span class="at">-o</span> HIV/ <span class="at">--confidence</span> none</span>
<span id="cb7-3"><a href=""></a><span class="co"># Plot mashtree bootstrap output</span></span>
<span id="cb7-4"><a href=""></a><span class="ex">Rscript</span> create_tree_HIV.R <span class="at">-t</span> HIV/mashtree.bootstrap.dnd <span class="at">-a</span> HIV/annotation.csv <span class="at">-o</span> HIV/ <span class="at">--confidence</span> color</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="plots-1" class="slide level2">
<h2>Plots</h2>
<div style="position: relative; width: 80%; height: 500px; margin: auto; text-align:center;">
<figure class="fragment fade-out" data-fragment-index="1" style="position:absolute; top:0; left:0; width:100%; margin:0;">
<a href="images/mashtree_phylo_tree_large.png" target="_blank"> <img src="images/mashtree_phylo_tree_large.png" style="width:45%; height:auto;"> </a>
<figcaption style="font-size:0.5em; margin-top:0.5em;">
HIV phylogenetic tree - mashtree version
</figcaption>
</figure>
<figure class="fragment" data-fragment-index="1" style="position:absolute; top:0; left:0; width:100%; margin:0;">
<a href="images/mashtree.bootstrap_phylo_tree_large.png" target="_blank"> <img src="images/mashtree.bootstrap_phylo_tree_large.png" style="width:45%; height:auto;"> </a>
<figcaption style="font-size:0.5em; margin-top:0.5em;">
HIV phylogenetic tree - mashtree bootstrap version
</figcaption>
</figure>
</div>
</section>
<section id="practical-applications" class="slide level2">
<h2>Practical applications</h2>
<ul>
<li class="fragment">Large-scale phylogenomics</li>
<li class="fragment">Metagenomics (binning, phylogenetic placement)</li>
<li class="fragment">Assembly-free analysis from raw sequencing reads</li>
<li class="fragment">Genomic epidemiology (e.g.&nbsp;SARS-CoV-2, bacterial outbreaks)</li>
</ul>
</section>
<section id="limitations" class="slide level2">
<h2>Limitations</h2>
<ul>
<li class="fragment"><p>Loss of positional information</p></li>
<li class="fragment"><p>Sensitivity to k-mer size</p></li>
<li class="fragment"><p>k-mer homoplasy (shared k-mers not due to ancestry)</p></li>
<li class="fragment"><p>Reliance on distance measures instead of explicit evolutionary models</p></li>
</ul>
</section></section>
<section>
<section id="conclusion" class="title-slide slide level1 center">
<h1>Conclusion</h1>

</section>
<section id="future-perspectives" class="slide level2">
<h2>Future perspectives</h2>
<ul>
<li class="fragment"><p>Hybrid approaches (alignment + alignment-free)</p></li>
<li class="fragment"><p>Statistical modeling of k-mer data (machine learning on compositional features)</p></li>
<li class="fragment"><p>Graphs of life instead of trees (capturing HGT and reticulate evolution)</p></li>
<li class="fragment"><p>Integration with pangenomics, multi-omics, and AI-based frameworks</p></li>
</ul>
</section>
<section id="conclusion-1" class="slide level2">
<h2>Conclusion</h2>
<ul>
<li class="fragment"><p>Take-home message: alignment-based and alignment-free methods are <strong>complementary tools</strong></p></li>
<li class="fragment"><p>Open questions for discussion:</p>
<ul>
<li class="fragment">When is alignment no longer useful?</li>
<li class="fragment">How do parameter choices affect different biological questions?</li>
</ul></li>
</ul>
</section>
<section id="thank-you" class="slide level2">
<h2>Thank you!</h2>
<div style="text-align: center; margin-top: 30px;">
<p><strong>Matheus Pimenta</strong><br>
📧 omatheuspimenta@outlook.com<br>
ESALQ/USP</p>
</div>
</section>
<section id="recommended-readings" class="slide level2 smaller scrollable">
<h2>Recommended readings</h2>
<p><span class="citation" data-cites="zielezinski_alignment-free_2017">Zielezinski et al. (<a href="#/recommended-readings" role="doc-biblioref" onclick="">2017</a>)</span><br>
<span class="citation" data-cites="zielezinski_benchmarking_2019">Zielezinski et al. (<a href="#/recommended-readings" role="doc-biblioref" onclick="">2019</a>)</span></p>

<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-zielezinski_benchmarking_2019" class="csl-entry" role="listitem">
Zielezinski, Andrzej, Hani Z. Girgis, Guillaume Bernard, Chris-Andre Leimeister, Kujin Tang, Thomas Dencker, Anna Katharina Lau, et al. 2019. <span>“Benchmarking of Alignment-Free Sequence Comparison Methods.”</span> <em>Genome Biology</em> 20 (1): 144. <a href="https://doi.org/10.1186/s13059-019-1755-7">https://doi.org/10.1186/s13059-019-1755-7</a>.
</div>
<div id="ref-zielezinski_alignment-free_2017" class="csl-entry" role="listitem">
Zielezinski, Andrzej, Susana Vinga, Jonas Almeida, and Wojciech M. Karlowski. 2017. <span>“Alignment-Free Sequence Comparison: Benefits, Applications, and Tools.”</span> <em>Genome Biology</em> 18 (1): 186. <a href="https://doi.org/10.1186/s13059-017-1319-7">https://doi.org/10.1186/s13059-017-1319-7</a>.
</div>
</div>
</section></section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<p><img src="images/logo.jpg" class="slide-logo"></p>
<div class="footer footer-default">
<p>Phylogenetics - Alignment-free Methods | Matheus Pimenta</p>
</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="index_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="index_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="index_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="index_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="index_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="index_files/libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="index_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="index_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="index_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="index_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="index_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': false,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"8\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: true,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'slide',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'fade',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        // Bounds for smallest/largest possible scale to apply to content
        minScale: 0.2,

        maxScale: 2,

        math: {
          mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
          const outerScaffold = trigger.parentElement.cloneNode(true);
          const codeEl = outerScaffold.querySelector('code');
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp('/' + window.location.host + '/');
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    

</body></html>